File: Algorithm/LatestTimebyReplacingHiddenDigits.cpp
// // Solution 1: Brute Force
// class Solution {
// public:
//     string maximumTime(string time) {
//         for (int m = 60 * 24 - 1; m >= 0; --m) {
//             int hh = m / 60;
//             int mm = m % 60;
//             if (time[0] != '?' && time[0] - '0' != hh / 10) continue;
//             if (time[1] != '?' && time[1] - '0' != hh % 10) continue;
//             if (time[3] != '?' && time[3] - '0' != mm / 10) continue;
//             if (time[4] != '?' && time[4] - '0' != mm % 10) continue;
//             time[0] = (hh / 10) + '0';
//             time[1] = (hh % 10) + '0';
//             time[3] = (mm / 10) + '0';
//             time[4] = (mm % 10) + '0';
//             break;
//         }
//         return time;
//     }
// };

// Solution 2: Rules
class Solution {
public:
    string maximumTime(string time) {
        if (time[0] == '?') time[0] = time[1] >= '4' && time[1] <= '9' ? '1' : '2';
        if (time[1] == '?') time[1] = time[0] == '2' ? '3' : '9';
        if (time[3] == '?') time[3] = '5';
        if (time[4] == '?') time[4] = '9';
        return time;
    }
};

File: Algorithm/LatestTimebyReplacingHiddenDigits.java
// // Solution 1: Brute Force
// // Enumerate all possible clock in reverse order and find the first matching one.
// // Time complexity: O(1)
// // Space complexity: O(1)
// class Solution {
//     public String maximumTime(String time) {
//         char[] ans = time.toCharArray();
//         for (int m = 24 * 60 - 1; m >= 0; --m) {
//             int hh = m / 60;
//             int mm = m % 60;
//             if (ans[0] != '?' && ans[0] - '0' != hh / 10) continue;
//             if (ans[1] != '?' && ans[1] - '0' != hh % 10) continue;
//             if (ans[3] != '?' && ans[3] - '0' != mm / 10) continue;
//             if (ans[4] != '?' && ans[4] - '0' != mm % 10) continue;
//             ans[0] = (char)(hh / 10 + '0');
//             ans[1] = (char)(hh % 10 + '0');
//             ans[3] = (char)(mm / 10 + '0');
//             ans[4] = (char)(mm % 10 + '0');
//         }
//         // return String.valueOf(ans); // char[]这样转成String也是对的。
//         return new String(ans);
//     }
// }

// Solution 2: Rules
// Using rules, fill from left to right.
// Time complexity: O(1)
// Space cmplexity: O(1)
class Solution {
    public String maximumTime(String time) {
        char[] ans = time.toCharArray();
        if (ans[0] == '?') ans[0] = ans[1] >= '4' && ans[1] <= '9' ? '1' : '2';
        if (ans[1] == '?') ans[1] = ans[0] == '2' ? '3' : '9';
        if (ans[3] == '?') ans[3] = '5';
        if (ans[4] == '?') ans[4] = '9';
        return String.valueOf(ans);
    }
}
